File: backend/apps/outputs/versions.py
"""Per-app version history, stored the way git stores history: content-addressed.

Each unique file's bytes are written ONCE to a per-app blob store (sha256 name,
zlib-compressed); a version is a tiny manifest mapping path -> blob digest. So a
run that changes one file out of fifty costs one new blob, not fifty, and 500
versions of a barely-changing app stay small. That keeps this feature from
making OpenSwarm feel heavier than it already is. We reuse the .swarm app
serializer (captures flat-inline AND webapp_template workspace apps; skips
node_modules/.venv/dist/.git, excludes .env), but never .swarm's pack() (it
refuses on secret-shaped fields, and a local snapshot must never decline to save
the user's own app). No git binary: it isn't guaranteed on a packaged Mac/Win."""
from __future__ import annotations

import hashlib
import json
import logging
import os
import shutil
import zlib
from datetime import datetime
from uuid import uuid4

from backend.apps.outputs.models import Output, OutputVersion
from backend.apps.outputs.workspace_io import _WALK_SKIP_DIRS, _save, load_output
from backend.apps.swarm.entities.apps import AppExportable
from backend.config.paths import OUTPUTS_VERSIONS_DIR, OUTPUTS_WORKSPACE_DIR

logger = logging.getLogger(__name__)

_MAX_FILE_BYTES = 25 * 1024 * 1024  # don't snapshot giant build artifacts


class _NullExportCtx:
    """serialize() wants an ExportContext but apps have no cross-refs to rewrite."""

    def bundle_id_for(self, etype, local_id):  # noqa: ARG002
        return None


_NULL_CTX = _NullExportCtx()


def _app_dir(output_id: str) -> str:
    return os.path.join(OUTPUTS_VERSIONS_DIR, output_id)


def _blobs_dir(output_id: str) -> str:
    return os.path.join(_app_dir(output_id), "blobs")


def _manifests_dir(output_id: str) -> str:
    return os.path.join(_app_dir(output_id), "manifests")


def _digest(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def _safe_join(folder: str, rel: str) -> str:
    dest = os.path.realpath(os.path.join(folder, rel))
    root = os.path.realpath(folder)
    if dest != root and not dest.startswith(root + os.sep):
        raise ValueError("version file path escapes the workspace")
    return dest


def _write_blob(output_id: str, data: bytes, digest: str) -> None:
    """Store bytes once, content-addressed. A blob that already exists is the
    whole point: that's a file unchanged since an earlier version, stored zero
    extra times."""
    path = os.path.join(_blobs_dir(output_id), digest)
    if os.path.exists(path):
        return
    os.makedirs(_blobs_dir(output_id), exist_ok=True)
    tmp = path + ".tmp"
    with open(tmp, "wb") as f:
        f.write(zlib.compress(data, 6))
    os.replace(tmp, path)


def _read_blob(output_id: str, digest: str) -> bytes | None:
    try:
        with open(os.path.join(_blobs_dir(output_id), digest), "rb") as f:
            return zlib.decompress(f.read())
    except (OSError, zlib.error):
        return None


def _read_manifest(output_id: str, version_id: str) -> dict | None:
    try:
        with open(os.path.join(_manifests_dir(output_id), f"{version_id}.json"), encoding="utf-8") as f:
            return json.load(f)
    except (OSError, json.JSONDecodeError, UnicodeDecodeError):
        return None


def _latest_manifest(output_id: str) -> dict | None:
    """Newest manifest by write time; only read for the dedupe check so capture
    stays O(1) reads rather than scanning every version's content."""
    folder = _manifests_dir(output_id)
    if not os.path.isdir(folder):
        return None
    js = [os.path.join(folder, f) for f in os.listdir(folder) if f.endswith(".json")]
    if not js:
        return None
    newest = max(js, key=os.path.getmtime)
    try:
        with open(newest, encoding="utf-8") as f:
            return json.load(f)
    except (OSError, json.JSONDecodeError, UnicodeDecodeError):
        return None


def _tree_hash(app_meta: dict, file_map: dict[str, str]) -> str:
    """Dedupe key over the snapshot's content: app metadata + path->digest map.
    Cheap because the per-file hashing already happened to name the blobs."""
    h = hashlib.sha256()
    h.update(json.dumps(app_meta, sort_keys=True).encode("utf-8"))
    h.update(b"\0")
    for path in sorted(file_map):
        h.update(path.encode("utf-8"))
        h.update(b"\0")
        h.update(file_map[path].encode("utf-8"))
        h.update(b"\0")
    return h.hexdigest()


def _snapshot(output: Output) -> tuple[dict, dict[str, bytes]]:
    app = AppExportable(output)
    return app.serialize(_NULL_CTX), app.files()  # files keyed workspace/<rel>


def list_versions(output_id: str) -> list[OutputVersion]:
    folder = _manifests_dir(output_id)
    if not os.path.isdir(folder):
        return []
    out: list[OutputVersion] = []
    for fname in os.listdir(folder):
        if not fname.endswith(".json"):
            continue
        m = _read_manifest(output_id, fname[:-5])
        if m is None:
            continue
        try:
            out.append(OutputVersion(**m))  # ignores the heavy keys (app_meta, files)
        except Exception:
            logger.warning("skipping unreadable version manifest %s", fname)
    out.sort(key=lambda v: v.created_at, reverse=True)
    return out


def capture(
    output_id: str,
    *,
    source: str = "auto",
    label: str = "",
    thumbnail: str | None = None,
) -> OutputVersion | None:
    """Snapshot current app state. Returns the existing latest version (no new
    manifest) when nothing changed, so unchanged runs don't pile up. None only if
    the app is gone."""
    output = load_output(output_id)
    if output is None:
        return None
    app_meta, files = _snapshot(output)

    file_map: dict[str, str] = {}
    for path, data in files.items():
        if len(data) > _MAX_FILE_BYTES:
            continue
        file_map[path] = _digest(data)
    tree = _tree_hash(app_meta, file_map)

    latest = _latest_manifest(output_id)
    parent_id = latest.get("id") if latest else None
    if latest is not None and latest.get("tree_hash") == tree:
        try:
            return OutputVersion(**latest)
        except Exception:
            pass  # corrupt latest: fall through and write a fresh manifest

    for path, data in files.items():
        d = file_map.get(path)
        if d is not None:
            _write_blob(output_id, data, d)

    vid = uuid4().hex
    manifest = {
        "id": vid,
        "created_at": datetime.now().isoformat(),
        "label": label or "",
        "source": source,
        "parent_id": parent_id,
        "thumbnail": thumbnail if thumbnail is not None else output.thumbnail,
        "tree_hash": tree,
        "app_meta": app_meta,
        "files": file_map,
    }
    os.makedirs(_manifests_dir(output_id), exist_ok=True)
    dest = os.path.join(_manifests_dir(output_id), f"{vid}.json")
    tmp = dest + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(manifest, f)
    os.replace(tmp, dest)
    return OutputVersion(**manifest)


def _restore_workspace(output_id: str, workspace_id: str, file_map: dict[str, str]) -> None:
    """Make the workspace match the snapshot. Deletes current authored files not
    in it, then writes the rest FROM BLOBS, skipping any file already byte-correct
    (so a restore writes only the diff). Keeps the live .env and build/cache dirs.
    Per-file writes: a crash mid-restore leaves a mixed tree, but the pre_restore
    backup restore() takes first is the real safety net."""
    folder = os.path.join(OUTPUTS_WORKSPACE_DIR, workspace_id)
    os.makedirs(folder, exist_ok=True)
    targets: dict[str, str] = {
        key[len("workspace/"):]: dig for key, dig in file_map.items() if key.startswith("workspace/")
    }

    for root, dirs, fnames in os.walk(folder):
        dirs[:] = [d for d in dirs if d not in _WALK_SKIP_DIRS]
        for fname in fnames:
            if fname == ".env":
                continue
            full = os.path.join(root, fname)
            rel = os.path.relpath(full, folder).replace(os.sep, "/")
            if rel not in targets:
                try:
                    os.remove(full)
                except OSError:
                    pass

    for rel, digest in targets.items():
        dest = _safe_join(folder, rel)
        if os.path.exists(dest):
            try:
                with open(dest, "rb") as f:
                    if _digest(f.read()) == digest:
                        continue  # already correct: don't rewrite
            except OSError:
                pass
        data = _read_blob(output_id, digest)
        if data is None:
            continue
        os.makedirs(os.path.dirname(dest), exist_ok=True)
        with open(dest, "wb") as f:
            f.write(data)


def restore(output_id: str, version_id: str) -> Output | None:
    """Bring the app back to an earlier version, in place. Saves the current
    state as a pre_restore version first so this is always undoable."""
    output = load_output(output_id)
    if output is None:
        return None
    manifest = _read_manifest(output_id, version_id)
    if manifest is None:
        return None

    target_label = manifest.get("label") or "an earlier version"
    capture(output_id, source="pre_restore", label=f"Before restoring '{target_label}'")

    app_meta = manifest.get("app_meta") or {}
    output.name = app_meta.get("name", output.name)
    output.description = app_meta.get("description", output.description)
    output.icon = app_meta.get("icon", output.icon)
    # Presence, not truthiness: a snapshot's empty {} schema must restore as empty.
    if app_meta.get("input_schema") is not None:
        output.input_schema = app_meta["input_schema"]
    output.files = app_meta.get("files") or {}
    if manifest.get("thumbnail") is not None:
        output.thumbnail = manifest["thumbnail"]
    now = datetime.now().isoformat()
    output.updated_at = now
    output.preview_updated_at = now

    if output.workspace_id:
        _restore_workspace(output_id, output.workspace_id, manifest.get("files") or {})
    _save(output)
    return output


def branch(output_id: str, version_id: str) -> str | None:
    """Make a brand-new app from an earlier version. Reuses the .swarm importer,
    which mints a fresh output id + workspace id and localizes a fresh .env."""
    manifest = _read_manifest(output_id, version_id)
    if manifest is None:
        return None
    app_meta = dict(manifest.get("app_meta") or {})
    app_meta["name"] = f"{app_meta.get('name') or 'App'} (copy)"
    files: dict[str, bytes] = {}
    for path, digest in (manifest.get("files") or {}).items():
        data = _read_blob(output_id, digest)
        if data is not None:
            files[path] = data
    from backend.apps.swarm.exportable import RemapTable
    try:
        return AppExportable.import_(app_meta, files, RemapTable())
    except Exception:
        # A partial import leaves an orphan workspace dir (small); better than a 500.
        logger.exception("branch import failed for %s/%s", output_id, version_id)
        return None


def delete_all(output_id: str) -> None:
    shutil.rmtree(_app_dir(output_id), ignore_errors=True)

File: engine/src/apps/outputs/versions.ts
// engine/src/apps/outputs/versions.ts -- SUB-5, a full TypeScript port of backend/apps/outputs/
// versions.py: per-app version history, stored the way git stores history: content-addressed.
//
// Each unique file's bytes are written ONCE to a per-app blob store (sha256 name, zlib-compressed);
// a version is a tiny manifest mapping path -> blob digest. So a run that changes one file out of
// fifty costs one new blob, not fifty. Reuses AppExportable (SUB-5's own real port, see that
// file's header) as the serializer -- captures both flat-inline AND webapp_template workspace
// apps, skips node_modules/.venv/dist/.git, excludes .env -- but never AppExportable/ziputil's
// secret-shaped-field refusal path: a local snapshot must never decline to save the user's own
// app. No git binary: it isn't guaranteed on a packaged Mac/Win host.

import { createHash } from 'node:crypto';
import { deflateSync, inflateSync } from 'node:zlib';
import {
  existsSync, mkdirSync, readFileSync, readdirSync, rmSync, statSync, writeFileSync, renameSync, unlinkSync,
} from 'node:fs';
import { dirname, join, relative, resolve as pathResolve, sep } from 'node:path';
import { randomUUID } from 'node:crypto';
import { AppExportable } from '../swarm/entities/appExportable';
import type { ExportContext } from '../swarm/exportable';
import { RemapTable } from '../swarm/exportable';
import { hydrateOutputVersion, type Output, type OutputVersion } from './models';
import { WALK_SKIP_DIRS, loadOutput, save } from './workspaceIo';
import { outputsVersionsDir, outputsWorkspaceDir } from './paths';

const P_MAX_FILE_BYTES = 25 * 1024 * 1024; // don't snapshot giant build artifacts

// serialize() wants an ExportContext but apps have no cross-refs to rewrite.
const P_NULL_CTX: ExportContext = { bundleIdFor: () => null };

function pAppDir(outputId: string, env: NodeJS.ProcessEnv): string {
  return join(outputsVersionsDir(env), outputId);
}

function blobsDir(outputId: string, env: NodeJS.ProcessEnv): string {
  return join(pAppDir(outputId, env), 'blobs');
}

function pManifestsDir(outputId: string, env: NodeJS.ProcessEnv): string {
  return join(pAppDir(outputId, env), 'manifests');
}

function pDigest(data: Buffer): string {
  return createHash('sha256').update(data).digest('hex');
}

function pSafeJoin(folder: string, rel: string): string {
  const dest = pathResolve(join(folder, rel));
  const root = pathResolve(folder);
  if (dest !== root && !dest.startsWith(root + sep)) throw new Error('version file path escapes the workspace');
  return dest;
}

function pWriteBlob(outputId: string, data: Buffer, digest: string, env: NodeJS.ProcessEnv): void {
  const dir = blobsDir(outputId, env);
  const path = join(dir, digest);
  if (existsSync(path)) return; // already stored: a file unchanged since an earlier version
  mkdirSync(dir, { recursive: true });
  const tmp = `${path}.tmp`;
  writeFileSync(tmp, deflateSync(data));
  renameSync(tmp, path);
}

function pReadBlob(outputId: string, digest: string, env: NodeJS.ProcessEnv): Buffer | null {
  try {
    return inflateSync(readFileSync(join(blobsDir(outputId, env), digest)));
  } catch {
    return null;
  }
}

interface Manifest {
  id: string;
  created_at: string;
  label: string;
  source: string;
  parent_id: string | null;
  thumbnail: string | null;
  tree_hash: string;
  app_meta: Record<string, unknown>;
  files: Record<string, string>; // rel path -> blob digest
}

export function readManifest(outputId: string, versionId: string, env: NodeJS.ProcessEnv = process.env): Manifest | null {
  try {
    return JSON.parse(readFileSync(join(pManifestsDir(outputId, env), `${versionId}.json`), 'utf8')) as Manifest;
  } catch {
    return null;
  }
}

/** Newest manifest by write time; only read for the dedupe check so capture stays O(1) reads
 * rather than scanning every version's content. */
function pLatestManifest(outputId: string, env: NodeJS.ProcessEnv): Manifest | null {
  const dir = pManifestsDir(outputId, env);
  if (!existsSync(dir) || !statSync(dir).isDirectory()) return null;
  const jsonFiles = readdirSync(dir).filter((f) => f.endsWith('.json'));
  if (jsonFiles.length === 0) return null;
  let newest = jsonFiles[0];
  let newestMtime = statSync(join(dir, newest)).mtimeMs;
  for (const f of jsonFiles.slice(1)) {
    const mtime = statSync(join(dir, f)).mtimeMs;
    if (mtime > newestMtime) {
      newest = f;
      newestMtime = mtime;
    }
  }
  try {
    return JSON.parse(readFileSync(join(dir, newest), 'utf8')) as Manifest;
  } catch {
    return null;
  }
}

/** Deterministic, recursively key-sorted JSON -- mirrors Python's `json.dumps(obj, sort_keys=True)`
 * (which sorts nested dict keys too, not just the top level). */
function pStableJson(value: unknown): string {
  if (Array.isArray(value)) return `[${value.map(pStableJson).join(',')}]`;
  if (value !== null && typeof value === 'object') {
    const keys = Object.keys(value as Record<string, unknown>).sort();
    return `{${keys.map((k) => `${JSON.stringify(k)}:${pStableJson((value as Record<string, unknown>)[k])}`).join(',')}}`;
  }
  return JSON.stringify(value);
}

/** Dedupe key over the snapshot's content: app metadata + path->digest map. Cheap because the
 * per-file hashing already happened to name the blobs. */
function pTreeHash(appMeta: Record<string, unknown>, fileMap: Record<string, string>): string {
  const h = createHash('sha256');
  h.update(pStableJson(appMeta));
  h.update('\0');
  for (const path of Object.keys(fileMap).sort()) {
    h.update(path);
    h.update('\0');
    h.update(fileMap[path]);
    h.update('\0');
  }
  return h.digest('hex');
}

function pSnapshot(output: Output): { appMeta: Record<string, unknown>; files: Record<string, Buffer> } {
  const app = new AppExportable(output);
  return { appMeta: app.serialize(P_NULL_CTX) as Record<string, unknown>, files: app.files() };
}

export function listVersions(outputId: string, env: NodeJS.ProcessEnv = process.env): OutputVersion[] {
  const dir = pManifestsDir(outputId, env);
  if (!existsSync(dir) || !statSync(dir).isDirectory()) return [];
  const out: OutputVersion[] = [];
  for (const fname of readdirSync(dir)) {
    if (!fname.endsWith('.json')) continue;
    const m = readManifest(outputId, fname.slice(0, -'.json'.length), env);
    if (m === null) continue;
    try {
      out.push(hydrateOutputVersion(m as unknown as Record<string, unknown>));
    } catch {
      // Skip an unreadable manifest rather than fail the whole list.
    }
  }
  out.sort((a, b) => (a.created_at < b.created_at ? 1 : a.created_at > b.created_at ? -1 : 0));
  return out;
}

export interface CaptureOptions {
  source?: 'auto' | 'manual' | 'pre_restore';
  label?: string;
  thumbnail?: string | null;
}

/** Snapshot current app state. Returns the existing latest version (no new manifest) when nothing
 * changed, so unchanged runs don't pile up. null only if the app is gone. */
export function capture(outputId: string, opts: CaptureOptions = {}, env: NodeJS.ProcessEnv = process.env): OutputVersion | null {
  const output = loadOutput(outputId, env);
  if (output === null) return null;
  const { appMeta, files } = pSnapshot(output);

  const fileMap: Record<string, string> = {};
  for (const [path, data] of Object.entries(files)) {
    if (data.length > P_MAX_FILE_BYTES) continue;
    fileMap[path] = pDigest(data);
  }
  const tree = pTreeHash(appMeta, fileMap);

  const latest = pLatestManifest(outputId, env);
  const parentId = latest?.id ?? null;
  if (latest !== null && latest.tree_hash === tree) {
    try {
      return hydrateOutputVersion(latest as unknown as Record<string, unknown>);
    } catch {
      // Corrupt latest: fall through and write a fresh manifest.
    }
  }

  for (const [path, data] of Object.entries(files)) {
    const digest = fileMap[path];
    if (digest !== undefined) pWriteBlob(outputId, data, digest, env);
  }

  const vid = randomUUID().replace(/-/g, '');
  const manifest: Manifest = {
    id: vid,
    created_at: new Date().toISOString(),
    label: opts.label ?? '',
    source: opts.source ?? 'auto',
    parent_id: parentId,
    // Python's own check is `thumbnail is not None` (not "was the kwarg passed"), so an explicit
    // null here falls through to output.thumbnail exactly like an omitted one would.
    thumbnail: opts.thumbnail !== undefined && opts.thumbnail !== null ? opts.thumbnail : output.thumbnail,
    tree_hash: tree,
    app_meta: appMeta,
    files: fileMap,
  };
  const dir = pManifestsDir(outputId, env);
  mkdirSync(dir, { recursive: true });
  const dest = join(dir, `${vid}.json`);
  const tmp = `${dest}.tmp`;
  writeFileSync(tmp, JSON.stringify(manifest));
  renameSync(tmp, dest);
  return hydrateOutputVersion(manifest as unknown as Record<string, unknown>);
}

/** Make the workspace match the snapshot. Deletes current authored files not in it, then writes
 * the rest FROM BLOBS, skipping any file already byte-correct (so a restore writes only the
 * diff). Keeps the live .env and build/cache dirs. Per-file writes: a crash mid-restore leaves a
 * mixed tree, but the pre_restore backup restore() takes first is the real safety net. */
function pRestoreWorkspace(outputId: string, workspaceId: string, fileMap: Record<string, string>, env: NodeJS.ProcessEnv): void {
  const folder = join(outputsWorkspaceDir(env), workspaceId);
  mkdirSync(folder, { recursive: true });
  const targets: Record<string, string> = {};
  for (const [key, digest] of Object.entries(fileMap)) {
    if (key.startsWith('workspace/')) targets[key.slice('workspace/'.length)] = digest;
  }

  const walkPrune = (dir: string): void => {
    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      const full = join(dir, entry.name);
      if (entry.isDirectory()) {
        if (WALK_SKIP_DIRS.has(entry.name)) continue;
        walkPrune(full);
        continue;
      }
      if (entry.name === '.env') continue;
      const rel = relative(folder, full).split(sep).join('/');
      if (!(rel in targets)) {
        try {
          unlinkSync(full);
        } catch {
          // Best-effort.
        }
      }
    }
  };
  walkPrune(folder);

  for (const [rel, digest] of Object.entries(targets)) {
    const dest = pSafeJoin(folder, rel);
    if (existsSync(dest)) {
      try {
        if (pDigest(readFileSync(dest)) === digest) continue; // already correct
      } catch {
        // Fall through to (re)write.
      }
    }
    const data = pReadBlob(outputId, digest, env);
    if (data === null) continue;
    mkdirSync(dirname(dest), { recursive: true });
    writeFileSync(dest, data);
  }
}

/** Bring the app back to an earlier version, in place. Saves the current state as a pre_restore
 * version first so this is always undoable. */
export function restore(outputId: string, versionId: string, env: NodeJS.ProcessEnv = process.env): Output | null {
  const output = loadOutput(outputId, env);
  if (output === null) return null;
  const manifest = readManifest(outputId, versionId, env);
  if (manifest === null) return null;

  const targetLabel = manifest.label || 'an earlier version';
  capture(outputId, { source: 'pre_restore', label: `Before restoring '${targetLabel}'` }, env);

  const appMeta = manifest.app_meta ?? {};
  output.name = (appMeta.name as string) ?? output.name;
  output.description = (appMeta.description as string) ?? output.description;
  output.icon = (appMeta.icon as string) ?? output.icon;
  // Presence, not truthiness: a snapshot's empty {} schema must restore as empty.
  if (appMeta.input_schema !== undefined && appMeta.input_schema !== null) {
    output.input_schema = appMeta.input_schema as Output['input_schema'];
  }
  output.files = (appMeta.files as Record<string, string>) ?? {};
  if (manifest.thumbnail !== undefined && manifest.thumbnail !== null) output.thumbnail = manifest.thumbnail;
  const now = new Date().toISOString();
  output.updated_at = now;
  output.preview_updated_at = now;

  if (output.workspace_id) pRestoreWorkspace(outputId, output.workspace_id, manifest.files ?? {}, env);
  save(output, env);
  return output;
}

/** Make a brand-new app from an earlier version. Reuses AppExportable.import_, which mints a
 * fresh output id + workspace id and localizes a fresh .env. */
export function branch(outputId: string, versionId: string, env: NodeJS.ProcessEnv = process.env): string | null {
  const manifest = readManifest(outputId, versionId, env);
  if (manifest === null) return null;
  const appMeta: Record<string, unknown> = { ...(manifest.app_meta ?? {}) };
  appMeta.name = `${(appMeta.name as string) || 'App'} (copy)`;
  const files: Record<string, Buffer> = {};
  for (const [path, digest] of Object.entries(manifest.files ?? {})) {
    const data = pReadBlob(outputId, digest, env);
    if (data !== null) files[path] = data;
  }
  try {
    return AppExportable.import_(appMeta, files, new RemapTable());
  } catch (err) {
    // A partial import leaves an orphan workspace dir (small); better than a 500.
    console.error(`[outputs] branch import failed for ${outputId}/${versionId}:`, err);
    return null;
  }
}

export function deleteAll(outputId: string, env: NodeJS.ProcessEnv = process.env): void {
  try {
    rmSync(pAppDir(outputId, env), { recursive: true, force: true });
  } catch {
    // Best-effort, matches versions.py's own ignore_errors=True.
  }
}
